keyboards/handwired/drevogramr_qmk/drevogramr_qmk.c

```diff
@@ -0,0 +1,178 @@
+==========================================================================*/
+
+
+/* initial matrix state, no columns is active */
+void keyMatrixClear(void) {
+
+    for(int i = 0; i < MATRIX_COLS; i++) {
+        writePinHigh(MATRIX_PIN_REG_SHIFT);
+        writePinLow(MATRIX_PIN_REG_SHIFT);
+    }
+
+    writePinHigh(MATRIX_PIN_REG_STORE);
+    writePinLow(MATRIX_PIN_REG_STORE);
+
+}
+
+
+
+void keyMatrixInit(void) {
+
+    for (int i = 0; i < MATRIX_ROWS ; i++) {
+        setPinInputHigh(keyMatrixRowPins[i]);
+    }
+
+    setPinOutput(MATRIX_PIN_REG_DATA);
+    writePinLow(MATRIX_PIN_REG_DATA);
+
+    setPinOutput(MATRIX_PIN_REG_STORE);
+    writePinLow(MATRIX_PIN_REG_STORE);
+
+    setPinOutput(MATRIX_PIN_REG_SHIFT);
+    writePinLow(MATRIX_PIN_REG_SHIFT);
+
+    keyMatrixClear();
+
+}
+
+
+
+/*
+    setting first column active.
+    assuming that keyMatrixClear was called right before this or keyMatrixShift was called at least MATRIX_COLS times
+*/
+void keyMatrixReset(void) {
+
+    writePinLow(MATRIX_PIN_REG_DATA);
+    
+    writePinHigh(MATRIX_PIN_REG_SHIFT);
+    writePinLow(MATRIX_PIN_REG_SHIFT);
+
+    writePinHigh(MATRIX_PIN_REG_DATA);
+
+    writePinHigh(MATRIX_PIN_REG_STORE);
+    writePinLow(MATRIX_PIN_REG_STORE);
+
+}
+
+
+
+/* setting next column active */
+void keyMatrixShift(void) {
+    
+    writePinHigh(MATRIX_PIN_REG_SHIFT);
+    writePinLow(MATRIX_PIN_REG_SHIFT);
+    
+    writePinHigh(MATRIX_PIN_REG_STORE);
+    writePinLow(MATRIX_PIN_REG_STORE);
+}
+
+
+
+/* read rows states and return bits in right order */
+uint8_t keyMatrixGetColRowValues(void) {
+
+    uint8_t res = 0;
+
+    for (int r = 0; r < MATRIX_ROWS; r++) {
+        res |= ((readPin(keyMatrixRowPins[r]) & 1) ? 0 : 1) << ((MATRIX_ROWS - 1) - r);
+    }
+
+    return res;
+}
+
+
+/*==========================================================================
+============================================================================
+==========================================================================*/
+
+
+
+void matrix_init_custom(void) {  
+
+    memset(oldMatrix, 0, matrixArraySize);
+    setPinOutput(LED_GUI_LOCK_PIN);
+    keyMatrixInit();
+
+}
+
+void keyboard_post_init_user () {
+    writePin(LED_GUI_LOCK_PIN, keymap_config.no_gui ? LED_PIN_ON_STATE : LED_PIN_OFF_STATE);
+}
+
+
+
+bool matrix_scan_custom(matrix_row_t current_matrix[]) {
+
+    bool changed = false;
+    memset(current_matrix, 0, matrixArraySize);
+
+    keyMatrixReset();
+    
+    for (int c = 0; c < MATRIX_COLS; c++) {
+
+        uint8_t rowValues = keyMatrixGetColRowValues();
+
+        for (int r = 0; r < MATRIX_ROWS; r++) {
+            current_matrix[r] |= ((rowValues & (1 << r)) ? 1 : 0) << c;
+        }
+    
+        keyMatrixShift();
+    }
+
+    changed = memcmp(current_matrix, oldMatrix, matrixArraySize) != 0;
+    memcpy(oldMatrix, current_matrix, matrixArraySize);
+
+    return changed;
+
+}
+
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
```

Comment:
```suggestion
bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
    if (!process_record_user(keycode, record)) { return false; }
    if(!record->event.pressed) {
```
or
```suggestion
__attribute__((weak)) bool process_record_user(uint16_t keycode, keyrecord_t *record) {
```

Otherwise, this may not compile properly.